phd-tests/framework/src/test_vm/config.rs

```diff
@@ -172,7 +207,29 @@ impl<'dr> VmConfig<'dr> {
             })
             .context("serializing Propolis server config")?;
 
-        let DiskSource::Artifact(boot_artifact) = self.boot_disk.source else {
+        // The first disk in the boot list might not be the disk a test *actually* expects to boot.
+        //
+        // If there are multiple bootable disks in the boot order, we'll assume they're all
+        // the same guest OS kind. So look for `boot-disk` - if there isn't a disk named
+        // `boot-disk` then fall back to hoping the first disk in the boot orrder is a bootable
+        // disk, and if *that* isn't a bootable disk, maybe the first disk is.
+        //
+        // TODO: theoretically we might want to accept configuration of a specific guest OS adapter
+        // and avoid the guessing games. So far the above supports existing tests and makes them
+        // "Just Work", but a more complicated test may want more control here.
+        let boot_disk =
+            self.disks.iter().find(|d| d.name == "boot-disk")
+                .or_else(|| if let Some(boot_order) = self.boot_order.as_ref() {
+                    boot_order.first().and_then(|name| self.disks.iter().find(|d| &d.name == name))
+                } else {
+                    None
+                })
```

Comment:
rehashing discussion elsewhere: it is actually important that the explicit boot order be ignored, since we now have tests that put a non-bootable disk up at the front of the list and expect to fall through.

if we had better support for one test having different OS kinds across its disks, it might make more sense to have an explicit setter for the guest OS kind and stop doing the guessing here. but that adds its own complication and even with the `boot_order` tests i've managed to avoid needing it.

that leaves this in the unfortunate limbo of "local minima of weird", i think